### ISSUE TITLE
[BUG] PIN_MASK missing

### ISSUE THREAD

**Issue by reynozeros** [RRZE-HPC/likwid#454]
I ran into some problems trying to do performance analysis of our hybrind MPI/OpenMP code with likwid.
After I found out that, no matter what I do, all software threads are getting pinned on a single hw thread per MPI process, I checked the output of likwid-pin. There in the line "[pthread wrapper] PIN_MASK:" nothing coes afterwards.

```console
likwid-pin -V 3 -c 0,1,2,3 -s 0x7 ./run.py 
DEBUG - [hwloc_init_cpuInfo:355] HWLOC CpuInfo Family 6 Model 158 Stepping 13 Vendor 0x0 Part 0x0 isIntel 1 numHWThreads 16 activeHWThreads 16
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 0 Thread 0 Core 0 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 8 Thread 1 Core 0 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 1 Thread 0 Core 1 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 9 Thread 1 Core 1 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 2 Thread 0 Core 2 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 10 Thread 1 Core 2 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 3 Thread 0 Core 3 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 11 Thread 1 Core 3 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 4 Thread 0 Core 4 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 12 Thread 1 Core 4 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 5 Thread 0 Core 5 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 13 Thread 1 Core 5 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 6 Thread 0 Core 6 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 14 Thread 1 Core 6 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 7 Thread 0 Core 7 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_nodeTopology:564] HWLOC Thread Pool PU 15 Thread 1 Core 7 Die 0 Socket 0 inCpuSet 1
DEBUG - [hwloc_init_cacheTopology:785] HWLOC Cache Pool ID 0 Level 1 Size 32768 Threads 2
DEBUG - [hwloc_init_cacheTopology:785] HWLOC Cache Pool ID 1 Level 2 Size 262144 Threads 2
DEBUG - [hwloc_init_cacheTopology:785] HWLOC Cache Pool ID 2 Level 3 Size 16777216 Threads 16
DEBUG - [affinity_init:539] Affinity: Socket domains 1
DEBUG - [affinity_init:541] Affinity: CPU die domains 1
DEBUG - [affinity_init:546] Affinity: CPU cores per LLC 8
DEBUG - [affinity_init:549] Affinity: Cache domains 1
DEBUG - [affinity_init:553] Affinity: NUMA domains 1
DEBUG - [affinity_init:554] Affinity: All domains 5
DEBUG - [affinity_addNodeDomain:370] Affinity domain N: 16 HW threads on 8 cores
DEBUG - [affinity_addSocketDomain:401] Affinity domain S0: 16 HW threads on 8 cores
DEBUG - [affinity_addDieDomain:438] Affinity domain D0: 16 HW threads on 8 cores
DEBUG - [affinity_addCacheDomain:474] Affinity domain C0: 16 HW threads on 8 cores
DEBUG - [affinity_addMemoryDomain:504] Affinity domain M0: 16 HW threads on 8 cores
DEBUG - [create_lookups:290] T 0 T2C 0 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 1 T2C 1 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 2 T2C 2 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 3 T2C 3 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 4 T2C 4 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 5 T2C 5 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 6 T2C 6 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 7 T2C 7 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 8 T2C 0 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 9 T2C 1 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 10 T2C 2 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 11 T2C 3 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 12 T2C 4 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 13 T2C 5 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 14 T2C 6 T2S 0 T2D 0 T2LLC 0 T2M 0
DEBUG - [create_lookups:290] T 15 T2C 7 T2S 0 T2D 0 T2LLC 0 T2M 0
Evaluated CPU string to CPUs: 0,1,2,3
Running: ./run.py
Using 4 thread(s) (cpuset: 0xf)
[pthread wrapper] 
[pthread wrapper] MAIN -> 1
[pthread wrapper] PIN_MASK: 
[pthread wrapper] SKIP MASK: 0x7
        threadid 140478346245888 -> SKIP 
        threadid 140478337853184 -> SKIP 
        threadid 140478329460480 -> SKIP 
[pthread wrapper] 
[pthread wrapper] MAIN -> 1
[pthread wrapper] PIN_MASK: 
[pthread wrapper] SKIP MASK: 0x7
        threadid 139758183470848 -> SKIP 
        threadid 139758101591808 -> SKIP 
        threadid 139758043674368 -> SKIP 
Roundrobin placement triggered
        threadid 139757858096896 -> hwthread 1 - OK
        threadid 139757849704192 -> hwthread 1 - OK
        threadid 139757766047488 -> hwthread 1 - OK
1 hardware thread(s) seem to be shared by 4 software thread(s). Check OMP_NUM_THREADS, OMP_PLACES, or OMP_PROC_BIND setting
...
```

I istalled likwid like described in the quick install guide:

```
VERSION=stable
wget http://ftp.fau.de/pub/likwid/likwid-$VERSION.tar.gz
tar -xaf likwid-$VERSION.tar.gz
cd likwid-*
vi config.mk # configure build, e.g. change installation prefix and architecture flags
make
sudo make install # sudo required to install the access daemon with proper permissions
```

To reproduce the situation on our cluster (where the same problem occurs), I changed the access to "perf_event" and set the "perf_event_paranoid" value to 2.
My local machine runs on Ubuntu 20.04 (Linux 5.13.0-25-generic x86_64) and the cluster on CentOS 7 (Linux 3.10.0-1160.42.2.el7.x86_64 x86_64).

**Comment by TomTheBear** [comment]
Thanks for reporting

This is an interesting case because there is a list forwarded to the pinning library (in environment variable `LIKWID_PIN`) but it contains only hardware thread `1`, the one the main PID is pinned to. Moreover, hwthread `1` is in the middle of your selection `0,1,2,3`. Can you output all LIKWID environment variables early in your Python script (before starting threads)?

The `perf_event` mode is not relevant in that case, so I hope with the the outputs I can reproduce it on any node.

**Comment by reynozeros** [comment]
The output for the env variables containing "LIKWID" is
```
LIKWID_INCLUDE=/usr/local/include/
LIKWID_LIB=/usr/local/lib/
LIKWID_NO_ACCESS=1
LIKWID_HWTHREADS=0,1,2,3
LIKWID_SKIP=0x7
LIKWID_PIN=1
```

**Comment by TomTheBear** [comment]
Thanks for the output.

The list for `LIKWID_PIN` is created [here](https://github.com/RRZE-HPC/likwid/blob/master/src/applications/likwid-pin.lua#L269-L274). It's using the same list as `LIKWID_HWTHREADS` as input, just a different way of iterating over it (hw thread for main application should be last in the list).

Unfortunately, while I can reproduce it using the environment variables directly, I cannot reproduce it with `likwid-pin`.

Does it work if you do a manual `os.environ["LIKWID_PIN"] = "1,2,3,0"` early in your Python script?

Moreover, it seems that your Python code starts two executables under the hood (`MAIN -> 1` twice). This is probably bad as they will be pinned to the same hardware threads (if it works).

**Comment by reynozeros** [comment]
Setting ```LIKWID_PIN``` first thing after all imports in the Python script still yields the empty ```PIN_MASK``` with all the same results as above, with the exception that ```LIKWID_PIN=1,2,3,0``` is printed for the env variables later in the Python script.
Concerning the main threads, I have no idea where that comes from. Afaik, we use SWIG python bindings for our C++ classes, meaning that Python should not start any other executable and just calls class functions.

**Comment by TomTheBear** [comment]
This was a stupid idea of mine, sorry for your work. The environment variable is evaluated by a constructor that runs before any of your commands/Python calls can be executed.

So, the best option for debugging this, is in `likwid-pin` itself:
```
$ cp $(which likwid-pin) .
$ vi likwid-pin
# Add some print_stderr() calls at lines L269-L274
# Use likwid_hwthreads instead of cpu_list
# Try #cpu_list instead of likwid.tablelength(cpu_list)
$ ./likwid-pin -c 0,1,2,3 -s 0x7 run.py
```

Does it work with different applications (`cd $LIKWID_SRD/test; make streamGCC; likwid-pin -c 0,1,2,3 -V 3 ./streamGCC`)?

**Comment by reynozeros** [comment]
Inserting some prints in the section
```
if num_threads > 1 then
    local pinString = tostring(math.tointeger(likwid_hwthreads[2]))
    print_stderr("cpu_list: ", table.concat(likwid_hwthreads,","))
    print_stderr("pinString: ", pinString)
    for i=3,likwid.tablelength(likwid_hwthreads) do
        pinString = pinString .. "," .. tostring(math.tointeger(likwid_hwthreads[i]))
        print_stderr("pinString: ", pinString)
    end
    pinString = pinString .. "," .. tostring(math.tointeger(cpu_list[1]))
    print_stderr("pinString: ", pinString)
    likwid.setenv("LIKWID_PIN", pinString)
```
results in
```
cpu_list: 
0,1,2,3
pinString: 
1
pinString: 
1,2
pinString: 
1,2,3
pinString: 
1,2,3,0
Running: ./run.py
```
no matter if I use ```cpu_list``` or ```likwid_hwthreads```, with the pin mask still missing and the env output still being
```
LIKWID_INCLUDE=/usr/local/include/
LIKWID_LIB=/usr/local/lib/
LIKWID_NO_ACCESS=1
LIKWID_HWTHREADS=0,1,2,3
LIKWID_SKIP=0x7
LIKWID_PIN=1
```

Is this env variable touched after that code part? I also noticed, that when I use ```likwid-mpirun``` in my slurm script for sbatch, the environment variables are not carried over. ```likwid-mpirun``` has to start a separate script, that loads all required modules to set the env variables. No idea if that could be connected.

**Comment by reynozeros** [comment]
If I just do ```srun ./run.py``` in my sbatch script, all env variables are carried over and everything works. Just when using ```likwid-mpirun``` it doesn't.
I will try playing around with these slurm switches and look into those env variables a bit more tomorrow.
Sadly I don't think you would have access on our DLR clusters. afaik those two are just for DLR institutes.

**Comment by TomTheBear** [comment]
The behavior of `likwid-mpirun` is in line with your described problem. `likwid-mpirun` uses `likwid-pin` or `likwid-perfctr` under the hood. So, if `likwid-pin` does not work, `likwid-mpirun` will fail as well.

**Comment by reynozeros** [comment]
After some playing around I really have no clue anymore. I can set ```LIKWID_PIN``` to anything and ```os.getenv("LIKWID_PIN")``` has exactly that value directly afterwards, but when printing the env in python only the part of the string before the first ```,``` retains its value. The first comma and everything after gets truncated.
And it only happens with ```LIKWID_PIN```. If I set another new env variable to some random value (containing some commas ofc) it stays the full string in the env printed from python.

**Comment by TomTheBear** [comment]
I had an idea which might cause the behavior but it is more on the wild guessing side.

`likwid-pin` itself is just a frontend for a `LD_PRELOAD` library with some additional environment variable like `LIKWID_PIN`. The library contains a constructor which reads `LIKWID_PIN`, `LIKWID_SKIP`, `LIKWID_SILENT` and uses `strtok_r` to process the content of `LIKWID_PIN`. If the `getenv(LIKWID_PIN)` call does return a direct pointer to the string in the execution context's environment and the processing with `strtok_r` somehow manipulates that string, the effect could be the one you are seeing.

Could you please check if this helps:
```
$ git clone git@github.com:RRZE-HPC/likwid.git
$ cd likwid
$ patch -p1 < pinlib-strtok.patch
$ make PREFIX=$HOME/likwid-install ACCESSMODE=perf_event
$ make PREFIX=$HOME/likwid-install ACCESSMODE=perf_event install
$ export LD_LIBRARY_PATH=$HOME/likwid-install/lib:$LD_LIBRARY_PATH
$ $HOME/likwid-install/bin/likwid-mpirun ...
```
[pinlib-strtok.patch.txt](https://github.com/RRZE-HPC/likwid/files/8834075/pinlib-strtok.patch.txt)

**Comment by TomTheBear** [comment]
The LIKWID tarball contains a MPI pinning test, please run that and supply the output. Also the output of likwid-topology inside the job would help.
```
$ cd likwid-X.Y.Z/test
$ make test-mpi
$ likwid-mpirun -d -mpi slurm -nperdomain C:1:4 likwid-topology
$ likwid-mpirun -d -mpi slurm -nperdomain C:1:4 ./test-mpi
```

Have you tried a skip mask of `0x0`?

**Comment by reynozeros** [comment]
The mpi-test shows, that two of the 4 threads are being pinned on the same HW thread.

And I didn't try ```-s 0x0``` before, but it doesn't make a difference when I tried it now.

[test-mpi.out.txt](https://github.com/RRZE-HPC/likwid/files/8876358/test-mpi.out.txt)

**Comment by TomTheBear** [comment]
Based on the output, your skip mask should be `0x3`.

```
Rank 3 Thread 1 running on Node n10616 core 15/15 with pid 61402 and tid 61926
Rank 3 Thread 2 running on Node n10616 core 12/12 with pid 61402 and tid 62037
Rank 3 Thread 3 running on Node n10616 core 13/13 with pid 61402 and tid 62141
Rank 3 Thread 0 running on Node n10616 core 12/12 with pid 61402 and tid 61402
likwid-pin(61289) ---test-mpi(61402)-+-pstree(62162)
                                     |-{test-mpi}(61435)
                                     |-{test-mpi}(61601)
                                     |-{test-mpi}(61926)
                                     |-{test-mpi}(62037)
                                     `-{test-mpi}(62141)
```
When you compare the PIDs printed by the `pstree` output and the ones printed by the application, the PIDs `61435` and `61601` are not used for work but are some progress/shepherd threads by MPI and/or OpenMP. With a skip mask of `0x3` you skip the first two threads in the pinning phase.

**Comment by reynozeros** [comment]
```test-mpi``` works fine with ```-s 0x3```, but in my application I still get the same warning that 3 HW threads are shared by 4 SW threads. If it was the same problem as in ```test-mpi``` shouldn't it be that the warning would say 4 HW threads are shared by 6 SW threads?
The output is still the same as before with the exception of
```
DEBUG: Using skip option -s 0x3 to skip pinning of shepard threads
```

**Comment by reynozeros** [comment]
Well, I asked my colleagues, how it is implemented and most don't know too much about it. As far as I understood, mpi4py is somewhat used and some custom omp wrapper to allow for supposedly more efficient threading directives or something.
Also: might the slurm option ```--mpi=pmi2``` pose problems? (probably not, I tried it without that option in the lua script, same result)
Could ```WARN: Cannot extract OpenMP vendor from executable or commandline, assuming no OpenMP``` be a hint to the problem?

Running without likwid and
```
#SBATCH --ntasks-per-node=32
#SBATCH --cpus-per-task=4
...
srun run.py
```
does not result in that warning.

So what is different, when likwid starts the ```.likwidscript``` with srun, I don't quite understand.

**Comment by TomTheBear** [comment]
The `--mpi=pmi2` option is partly set by `likwid-mpirun` itself, so that's fine.

The `mpi` and `omp` check is very simple by calling `ldd` on the executable. This does not work in your case because `run.py` is a script, not a binary. But you can overwrite the setting with `-omp gnu` or similar:
```
-mpi <id>		 Specify which MPI should be used. Possible values: openmpi, intelmpi, slurm and mvapich2
			 If not set, module system is checked
-omp <id>		 Specify which OpenMP should be used. Possible values: gnu and intel
			 Only required for statically linked executables.
```

It might be (I havn't checked the code) that is refuses to use multiple threads if no OpenMP is detected. So try it with a manual setting for OpenMP.

The `.likwidscript` is basically:
```sh
if $RANK == 0; then 
    likwid-pin -c $RANK0_PINSTR $EXEC
else if $RANK == 1; then
    likwid-pin -c $RANK1_PINSTR $EXEC
[...]
fi
```

As a remark, `likwid-mpirun` is just a Lua script. You can copy it and manipulate it (similar to `likwid-pin` in one of my previous posts).

Did I get that right from your answer: nobody knows exactly how your application works?

**Comment by reynozeros** [comment]
Sorry for my late response, I was out with covid for a few weeks, but doing fine again.

The majority of devs for our application are at a different institute than mine, so I have to ask around who knows how it works.

Specifying ```-omp``` didn't make any difference except that the warning about the OpenMP vendor vanished.

I checked the runtime again and it seems to be fine, maybe I just looked at the wrong results last time. So the pinning seems to be done correctly.
Without the markers I also get similar results as when I use the application in pure mpi mode.
Only with the marker api i get this problem in hybrid mode, that only one thread per process has actual values for the counters (except for the number of region calls, there every thread has a value).
I also tried just using ```likwid-perfctr``` with ```-C 0-127 -s 0x7``` threads to see if the problem is also there and then I get the error message:
```
Line 5 0 124 3 2.081962e-04 0
 not a valid region values line
Line 5 0 125 3 3.885229e-04 0
 not a valid region values line
Line 5 0 126 3 2.825648e-04 0
 not a valid region values line
Line 5 0 127 3 3.471937e-04 0
 not a valid region values line
[...]/likwid_install/zen2/gcc-10.3.0/lua-5.3.5-dmgthb/bin/lua: ...zen2/gcc-10.3.0/likwid-5.2.1-alujem/share/lua/likwid.lua:834: bad argument #1 to 'for iterator' (table expected, got nil)
stack traceback:
        [C]: in function 'next'
        ...zen2/gcc-10.3.0/likwid-5.2.1-alujem/share/lua/likwid.lua:834: in function 'likwid.printOutput'
        ...l/zen2/gcc-10.3.0/likwid-5.2.1-alujem/bin/likwid-perfctr:1128: in main chunk
        [C]: in ?
```

The first few lines are the same many hundred times, probably once per thread per region or so.

I also checked if ```LIKWID_MARKER_THREADINIT``` is called in the parallel region and that seems to be the case, so I have no idea, whats the problem here.

Also interesting to note: The skip mask that works when I just use ```likwid-pin``` and also seems to work (no warning messages) with ```likwid-mpirun```, doesn't work here. I again get the warning, that 128 software threads seem to be mapped to 127 hw threads.

**Comment by reynozeros** [comment]
I'm not quite sure that that is truly my problem, as I was using ```likwid-perfctr``` just to try it out without mpi.
Th problem is that it runs fine and with plausible results without ```-m```, but with the markers I get this error message.

And I don't use ```LIKWID_MARKER_REGISTER()``` because I created some marker groups of different granularity, but not all groups are necessarily called, so i don't want to bloat my output with warnings of regions that got never entered.
So I don'T necessarily need ```LIKWID_MARKER_THREADINIT```? I will try without then. And I will also try again with ```LIKWID_MARKER_REGISTER()``` to see what happens.

**Comment by reynozeros** [comment]
I think this Issue should be closed, With yor provided patch, the ```PIN_MASK``` is not missing anymore. Is this patch included in the next release? When I tell my colleagues how to use likwid with our software, I would like to tell them just to install it and not having to manually patch it.
For my other Problems I will open another issue. It seems to be a different Problem.

**Comment by TomTheBear** [comment]
The patch will be in the upcoming release, yes. I just added it.

I also worked on the skip_masks, so maybe this is also fixed for you with the next release.
